File: src/BreakfastProvider.Api/Events/Outbox/OutboxWriter.cs
using System.Text.Json;
using BreakfastProvider.Api.Storage;
using Microsoft.Azure.Cosmos;

namespace BreakfastProvider.Api.Events.Outbox;

public class OutboxWriter(Container container) : IOutboxWriter
{
    private static readonly JsonSerializerOptions SerializerOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase
    };

    public async Task WriteAsync<TDocument, TEvent>(TDocument document, TEvent @event, string partitionKey, string destination, CancellationToken cancellationToken = default)
        where TDocument : class
        where TEvent : class
    {
        var eventType = typeof(TEvent).Name;
        var payload = JsonSerializer.Serialize(@event, SerializerOptions);

        var message = new OutboxMessage
        {
            PartitionKey = partitionKey,
            EventType = eventType,
            Destination = destination,
            Payload = payload,
            Status = OutboxMessageStatus.Pending,
            CreatedAt = DateTime.UtcNow
        };

        var batch = container.CreateTransactionalBatch(new PartitionKey(partitionKey));
        batch.CreateItem(document);
        batch.CreateItem(message);

        using var response = await batch.ExecuteAsync(cancellationToken);

        if (!response.IsSuccessStatusCode)
            throw new CosmosException(
                $"Transactional batch failed with status {response.StatusCode}",
                response.StatusCode, 0, response.ActivityId, response.RequestCharge);
    }
}

File: java/breakfast-provider-app/src/main/java/io/lemonlion/breakfast/events/outbox/OutboxWriter.java
package io.lemonlion.breakfast.events.outbox;

/**
 * Twin of C# {@code IOutboxWriter}. Atomically persists a domain document and its outbox message
 * (Cosmos transactional batch in the real impl; a single in-memory commit in the fake).
 */
public interface OutboxWriter {

    /**
     * @param document    the domain document to store
     * @param event       the event to serialize into the outbox message payload
     * @param partitionKey the shared partition key (document and message must co-locate for the batch)
     * @param destination one of {@link OutboxDestinations}
     */
    <TDocument, TEvent> void write(TDocument document, TEvent event, String partitionKey, String destination);
}
